apps/prairielearn/src/ee/lib/ai-grading.ts

```diff
@@ -21,13 +28,148 @@ import { getQuestionCourse } from '../../lib/question-variant.js';
 import { createServerJob } from '../../lib/server-jobs.js';
 import * as questionServers from '../../question-servers/index.js';
 
+import { createEmbedding, vectorToString } from './contextEmbeddings.js';
+
 const sql = loadSqlEquiv(import.meta.url);
 
 const SubmissionVariantSchema = z.object({
   variant: VariantSchema,
   submission: SubmissionSchema,
 });
 const GPTGradeSchema = z.object({ grade: z.number(), feedback: z.string() });
+const GradedExampleSchema = z.object({
+  submission_text: z.string(),
+  score_perc: z.number(),
+  instance_question_id: z.string(),
+});
+type GradedExample = z.infer<typeof GradedExampleSchema>;
+
+async function generateGPTPromptWithExamples({
+  question_prompt,
+  student_answer,
+  example_submissions,
+}: {
+  question_prompt: string;
+  student_answer: string;
+  example_submissions: GradedExample[];
+}): Promise<
+  (
+    | {
+        role: 'system';
+        content: string;
+      }
+    | {
+        role: 'user';
+        content: string;
+      }
+  )[]
+> {
+  const messages: (
+    | {
+        role: 'system';
+        content: string;
+      }
+    | {
+        role: 'user';
+        content: string;
+      }
+  )[] = [];
```

Comment:
Plus or minus some Prettier formatting:

```suggestion
  (
    | {
        role: 'system' | 'user';
        content: string;
      }
  )[]
> {
  const messages: (
    | {
        role: 'system' | 'user';
        content: string;
      }
  )[] = [];
```